File: Ch.2 - Solution of Equations in One Variable/bisection.py
#!/Users/devere/miniconda3/bin/python3
import os

"""
bisection.py: my implementation of the Bisection method
Author: Devere Anthony Weaver
"""
def f(x):
    """ function to be evaluated (must change with different function) """
    return ((x*x*x) + (4*x*x) - 10)

def bisection(a, b, t, n):
    """ function that implements bisection method algorithm """
    counter = 0
    fa = f(a)
    p = 0
    l = a
    u = b
    print(f"n\ta\t\t\tb\t\t\tp (approx.)\t\t\tf(p)\t\t\t\tError")
    print("__________________________________________________________________________________________________________________________________________")
    while counter < n:
        p = a + (b-a)/2
        fp = f(p)
        if (fp == 0) or ((b-a)/2 < t):    # if exact root found or error tolerance is reached
            counter += 1
            print(f"{counter:>02}\t{a:<16}\t{b:<11}\t\t{p:<16}\t\t{f(p):<16}\t\t{b-p:<16}")
            break
        elif (fa * fp > 0):    # same signs, shift right
            a = p
            fa = fp
        else:    # different signs, shift left
            b = p
        counter += 1
        print(f"{counter:>02}\t{a:<16}\t{b:<11}\t\t{p:<16}\t\t{f(p):<16}\t\t{b-a:<16}")
    print(f"\nAfter {counter} iterations, the approximation for the root in [{l},{u}] is ~{p}\nwith error {b-p}")
    
if __name__ == '__main__':
    while True:
        os.system('cls' if os.name == 'nt' else 'clear')
        print("The Bisection Method\n____________________")
        a = input("Enter a > ")
        b = input("Enter b > ")
        t = input("Enter error tolerance > ")
        n = input("Enter number of iterations > ")
        bisection(float(a), float(b), float(t), int(n))
        cont = input("\nContinue [y/N]? > ")
        if (cont.lower() != 'y'):
            print("\nGoodbye...")
            break
        else: 
            print("\n")
            continue

File: Ch.2 - Solution of Equations in One Variable/Bisection/bisection.cpp
/* bisection.cpp - my implementation of the Bisection method
 * currently being studied for Numerical Analysis
 *
 * NOTE: This only works for the given function defined in 
 * f(), in order to evaluate a different function, f's return 
 * value needs to be changed.
 * 
 * Author: Devere Anthony Weaver
 */

#include <iostream>
#include <cmath>
#include <string>
#include <iomanip>

double f(double x)    // Change return value to evaluate a different function
{
	return (pow(x, 3) + 4*pow(x, 2) - 10);
}

int main()
{
	double a{}, b{}, tol{}, p{};
	int n{}, counter{};
	double fa = f(a);
	
	while (true)
	{
		std::system("clear");
		std::cout << "The Bisection Method\n___________________" << std::endl;
		std::cout << "\nEnter a > ";
		std::cin >> a;
		std::cout << "Enter b > ";
		std::cin >> b;
		std::cout << "Enter number of iterations (n) > ";
		std::cin >> n;
		std::cout << "Enter error tolerance > ";
		std::cin >> tol;

		std::cout << std::setw(20) << "\nIteration" << std::setw(25) << "a"
		<< std::setw(20) << "b" << std::setw(20) << "p" << std::setw(20) << "f(p)"
		<< std::setw(20) << "Error"
		<< "\n___________________________________________________________________________________________________________________"
		<< std::endl;

		for (int i{0}; i < n; i++)
		{
			counter = i+1;
			p = (a+b)/2.;
			double fp = f(p);    // evaluate f at p

			std::cout << std::setw(2) << i + 1 << " | " << std::setw(30) << std::setprecision(10) <<  a 
				  << std::setw(20) << b 
				  << std::setw(20) << p 			  
				  << std::setw(20) << std::setprecision(5) << fp
				  << std::setw(20) << std::setprecision(10) << (b - p)
				  << "\n___________________________________________________________________________________________________________________"
				  << std::endl;

			if (fp == 0 || ((b-a)/2. < tol))    // exact solution found or tolerance reached
				break;   
			if (fa * fp > 0)    // same signs, shift right
			{
				a = p;
				fa = fp;
			}
			else
				b = p;    // different signs, shift left
		}

		std::cout << "\nAfter " << counter << " iterations, the approximation for the root is " 
			  << std::setprecision(10) << p << std::endl;

		std::string cont{};
		std::cout << "\nContinue [y/N] > ";
		std::cin >> cont;
		if (cont == "y" || cont == "Y")
			continue;
		else 
			break;
	} // end while
}

